### ISSUE TITLE
Give an ability to add another video streaming service

### ISSUE THREAD

**Issue by Joe23232** [SkyTubeTeam/SkyTube#946]
It would be nice if we can add more video streaming services of our own choice.

Currently services such as BitChute are growing due to YouTube's severe censorship and therefore alternative services such as BitChute are able to provide alternative content. Unfortunately New Pipe only shows contents from a limited sources such as YouTube, PeerTube etc. I wish that the app also allows users to add whatever platforms they would like to it.

Additionally it would be nice if there was a section where it would show all the desired content from different platforms (for example YouTube, BitChute etc).

**Comment by gzsombor** [comment]
If you want to add new streaming services, you can add them to the NewPipeExtractor library, what is used by SkyTube. After, you should implement a service selector in SkyTube, which maybe not that big, but there could be several approach. I mean, maybe having a separate 'subscription feed' tab for each service, or not, etc.

**Comment by Joe23232** [comment]
> If you want to add new streaming services, you can add them to the NewPipeExtractor library, what is used by SkyTube.

Since you said "NewPipeExtractor" is this also in the app "[NewPipe](https://newpipe.net/)"?

> After, you should implement a service selector in SkyTube, which maybe not that big, but there could be several approach. I mean, maybe having a separate 'subscription feed' tab for each service, or not, etc.

How hard is it to do this?

**Comment by gzsombor** [comment]
Yes, [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor)  is used by both NewPipe and SkyTube.

Well, how hard? It's hard to guess. It depends on how customizable you want to go. I mean, if you want an app, which only supports SoundCloud, but not YouTube, it's a couple of days work. If you want a way to add custom PeerTube servers, and the UI for the different services, etc, then it could be a couple of weeks of work. And of course, some thinking ahead to design these feature is also needed. And  maybe after a week of work, you notice, that it wont be too user friendly, or the implementation is too complicated, so you have to start it from scratch.

**Comment by Joe23232** [comment]
WHat about adding Bitchute then?

**Comment by gzsombor** [comment]
I guess, technically the same problem. What about it?

**Comment by Joe23232** [comment]
Like how difficult is it to add BitChute?

**Comment by gzsombor** [comment]
If you ask me, I don't know the answer. You can see, how big are similar service implementations in NewPipeExtractor - basically you need to parse the html of your service (or find working urls which could return the relevant information in JSON or XML). And do it for at least to cover:
* parse the 'trending'/'main' page for videos/channels
* parse the page for one video - to get the urls for the video streams
* parse a channel page for videos
* implement searching for videos/channels
